### PR TITLE
Hw 38 solution

### PR DESCRIPTION
**Pull request by ipantazi**

### CODE REVIEW

File: src/main/java/mate/academy/onlinebookstore/controller/BookController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/books")
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/controller/BookController.java]:
```suggestion
@RequestMapping("/books")
```

File: src/main/java/mate/academy/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public BookDto createBook(@RequestBody CreateBookRequestDto bookRequestDto) {
+        return bookService.save(bookRequestDto);
+    }
+
+    @PutMapping("/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public int updateBook(@PathVariable Long id,
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/controller/BookController.java]:
```suggestion
    public BookDto updateBook(@PathVariable Long id,
```
Better return updated book

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    private BigDecimal price;
+
+    @Column(nullable = false)
+    private String description;
+
+    @Column(name = "cover_image", nullable = false)
+    private String coverImage;
+
+    @Column(name = "is_deleted", nullable = false)
+    private boolean deleted = false;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/mate/academy/onlinebookstore/repository/BookRepository.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @Query(value = "UPDATE Book SET title = :title, author = :author, isbn = :isbn, "
+            + "price = :price, description = :description, coverImage = :coverImage WHERE id = :id "
+            + "AND deleted = false")
+    int updateBookById(@Param("id") Long id,
+                        @Param("title") String title,
+                        @Param("author") String author,
+                        @Param("isbn") String isbn,
+                        @Param("price") BigDecimal price,
+                        @Param("description") String description,
+                        @Param("coverImage") String coverImage);
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/repository/BookRepository.java]: Why do you need it? Can you use the method save CrudRepository?

File: src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public int update(Long id, CreateBookRequestDto bookRequestDto) {
+        return bookRepository.updateBookById(id,
+                bookRequestDto.getTitle(),
+                bookRequestDto.getAuthor(),
+                bookRequestDto.getIsbn(),
+                bookRequestDto.getPrice(),
+                bookRequestDto.getDescription(),
+                bookRequestDto.getCoverImage());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java]: Use Mapper for it;) Check if the book exists before updating

File: src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void deleteById(Long id) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java:45]: the same. Let’s check that the book exists by id.

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]: leave one empty line at the end of the file

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes ;)

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "books")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]: check if `@Where` is deprecated or not

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    @Column(nullable = false)
+    private String description;
+
+    @Column(name = "cover_image", nullable = false)
+    private String coverImage;
+
+    @Column(name = "is_deleted", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]:
```suggestion
    @Column(nullable = false)
    private String coverImage;

    @Column(nullable = false)
```

File: src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public BookDto findById(Long id) {
+        return bookMapper.toBookDto(findBookById(id));
+    }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto bookRequestDto) {
+        Book updatedBook = bookMapper.updateBookFromDto(bookRequestDto, findBookById(id));
+        bookRepository.save(updatedBook);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java]:
```suggestion
        Book book = findBookById(Long id);
        bookMapper.updateBookFromDto(bookRequestDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

File: src/main/java/mate/academy/onlinebookstore/mapper/BookMapper.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+public interface BookMapper {
+    BookDto toBookDto(Book book);
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "deleted", ignore = true)
+    Book toModel(CreateBookRequestDto bookRequestDto);
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "deleted", ignore = true)
+    Book updateBookFromDto(CreateBookRequestDto bookRequestDto, @MappingTarget Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/mapper/BookMapper.java]:
```suggestion
    void updateBookFromDto(CreateBookRequestDto bookRequestDto, @MappingTarget Book book);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    @Column(nullable = false, unique = true)
+    private String isbn;
+
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    @Column(nullable = false)
+    private String description;
+
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]:
```suggestion
```

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    private String author;
+
+    @Column(nullable = false, unique = true)
+    private String isbn;
+
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
two comments

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+              - column:
+                  name: description
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]:
```suggestion
```

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+              - column:
+                  name: price
+                  type: decimal
+                  constraints:
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
don't forget about changesets when you make some changes in entity

File: src/main/java/mate/academy/onlinebookstore/dto/BookDto.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package mate.academy.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@RequiredArgsConstructor
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/dto/BookDto.java:10]: Why do you need this annotation here?

File: src/main/java/mate/academy/onlinebookstore/dto/CreateBookRequestDto.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
+package mate.academy.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@RequiredArgsConstructor
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/onlinebookstore/dto/CreateBookRequestDto.java:10]: the same

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! A few minor comments;)

**Pull request closed without merging**